### PR TITLE
AT-59 - auto recognise filetype

### PR DESCRIPTION
**Pull request by benclayton98**
Co-authored-by: OllieVanD

https://dfacademy.atlassian.net/jira/software/projects/AT/boards/4?issueParent=10138&selectedIssue=AT-59

We removed the need for the user to input the file type that they were uploading and have now ensured that the program correctly identifies which file type has been uploaded on the server-side

### CODE REVIEW

**Review by grantwhiteman — DISMISSED** [review]

File: src/fileUpload/headerCheck.js [hunk 1]
```diff
@@ -6,32 +6,36 @@ const correctHeaders = {
...
-  }
-  else { 
-    headerCheck.validFile = false;
-    headerCheck.errors = failedHeaderCheck(headers);
-    return headerCheck;
-  }
-};
-
-function failedHeaderCheck (headers) {
   if (checkArraysAreEqual(headers, correctHeaders.moduleChallenge)) {
```
Comment (dearshrewdwit) [inline src/fileUpload/headerCheck.js:20]: Can we reduce this if statement to something like:

```js
const { checkArraysAreEqual } = require("./utility.js");
const correctHeaders = {
  selfAssessment: ["StudentId", "confidenceScore", "overallScore", "studentReason", "studentFeedback", "dueDate", "submissionDate"],
  moduleChallenge: ["StudentId", "challengeName", "language", "studentScore", "coachScore", "dueDate", "submissionDate"],
  student: ["firstName", "lastName", "githubUsername", "email", "gender", "background", "CohortId"]
};
const headerCheck = { 
  validFile: true,
  errors: "",
  fileType: ""
};
  
const headerChecker = (headers) => {
  const foundHeaders = Object.values(correctHeaders).find(correctSet => checkArraysAreEqual(correctSet, header))
  const fileType = Object.keys(correctHeaders).find(fileType => checkArraysAreEqual(correctHeaders[fileType], foundHeaders))
  headerCheck.fileType = fileType || ""
  if (fileType === undefined) headerCheck.validFile = false
  return headerCheck
}

module.exports = headerChecker
```

**Review by dearshrewdwit — CHANGES_REQUESTED** [review]
Nice. See refactoring opportunity

**Review by grantwhiteman — APPROVED** [review]

File: src/fileUpload/headerCheck.js [hunk 2]
```diff
@@ -1,37 +1,27 @@
...
-  if (checkArraysAreEqual(headers, correctHeaders[table])){
-    return headerCheck;
-  }
-  else { 
+  const foundHeaders = Object.values(correctHeaders).find(correctSet => checkArraysAreEqual(correctSet, headers));
+  if(foundHeaders === undefined) {
     headerCheck.validFile = false;
-    headerCheck.errors = failedHeaderCheck(headers);
+    headerCheck.errors = `Headers: [${headers}] does not match any valid headers`;
+    headerCheck.fileType = "";
```
Comment (dearshrewdwit) [inline src/fileUpload/headerCheck.js:18]: NB: It's already set as an empty string on line 12.

You can get around some of your problems here by returning false early from your checkArrays method if the args are not both arrays.

File: src/fileUpload/headerCheck.js [hunk 3]
```diff
@@ -1,37 +1,27 @@
...
-
 const correctHeaders = {
   selfAssessment: ["StudentId", "confidenceScore", "overallScore", "studentReason", "studentFeedback", "dueDate", "submissionDate"],
   moduleChallenge: ["StudentId", "challengeName", "language", "studentScore", "coachScore", "dueDate", "submissionDate"],
   student: ["firstName", "lastName", "githubUsername", "email", "gender", "background", "CohortId"]
 };
 
-module.exports = function headerChecker (headers, table) {
+const headerChecker = (headers) => {
   const headerCheck = {
```
Comment (dearshrewdwit) [inline src/fileUpload/headerCheck.js:9]: Is there a reason not to move this outside of function?

Comment (OllieVanD) [reply src/fileUpload/headerCheck.js:9]: If you save it outside of the function you'd then have to resave the values as default again inside the function as it seems to keep the values stored inside the browser memory when it is run again. Probably not a major issue with prod, but when running Cypress tests, if you assign a value in a previous test, then all future ones will have that same value until it is then reassigned.

File: src/fileUpload/headerCheck.js (hunk 2 above)
Comment (OllieVanD) [reply src/fileUpload/headerCheck.js:18]: Ah yeah forgot to change that back. 

Yeah that's definitely better.

File: src/fileUpload/headerCheck.js (hunk 3 above)
Comment (dearshrewdwit) [reply src/fileUpload/headerCheck.js:9]: Ok 👍

**Review by grantwhiteman — DISMISSED** [review]

File: src/fileUpload/headerCheck.js [hunk 4]
```diff
@@ -1,37 +1,26 @@
...
+    fileType: ""
   };
-  if (checkArraysAreEqual(headers, correctHeaders[table])){
-    return headerCheck;
-  }
-  else { 
+  const foundHeaders = Object.values(correctHeaders).find(correctSet => checkArraysAreEqual(correctSet, headers));
+  const fileType = Object.keys(correctHeaders).find(fileType => checkArraysAreEqual(correctHeaders[fileType], foundHeaders));
+  headerCheck.fileType = fileType || "";
+  console.log(fileType);
```
Comment (dearshrewdwit) [inline src/fileUpload/headerCheck.js:17]: remove debugging statement

File: src/fileUpload/fileUpload.js [hunk 5]
```diff
@@ -19,31 +19,25 @@ class FileUploader {
...
 
-  async process(fileType, csvFile) {
-    this.history = fileType;
-    this.table = this.dbCheck(fileType);
+  async process(csvFile) {
     this.data = this.fileConvertor(csvFile);
-    this.data.map((dataObject, i) => dataObject.counter = i + 1);
-    const headerCheck = headerChecker(this.headers, fileType);
-        
+    // console.log(this.headers)
```
Comment (dearshrewdwit) [inline src/fileUpload/fileUpload.js:24]: remove commented code

**Review by dearshrewdwit — CHANGES_REQUESTED** [review]
Small cleanup needed

File: src/fileUpload/utility.js [hunk 6]
```diff
@@ -1,6 +1,13 @@
 function checkArraysAreEqual (arr1, arr2) {
```
Comment (dearshrewdwit) [inline src/fileUpload/utility.js:1]: I don't imagine this has a set of tests - can you add a new issue for this?

**Review by dearshrewdwit — APPROVED** [review]
Remember to talk to your front end team - and let them know of this change. 

It shouldn't cause problems to send over the extra data in the request until the front end is updated.

**Review by OllieVanD — APPROVED** [review]

**Pull request merged**